sentry/src/main/java/io/sentry/instrumentation/file/SentryFileInputStream.java

```diff
@@ -0,0 +1,197 @@
+package io.sentry.instrumentation.file;
+
+import com.jakewharton.nopen.annotation.Open;
+import io.sentry.ISpan;
+import io.sentry.Sentry;
+import io.sentry.SpanStatus;
+import io.sentry.util.StringUtils;
+import java.io.File;
+import java.io.FileDescriptor;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * An implementation of {@link java.io.FileInputStream} that creates a {@link io.sentry.ISpan} for
+ * reading operation with filename and byte count set as description
+ *
+ * Note, that span is started when this InputStream is instantiated via constructor and finishes
+ * when the {@link java.io.FileInputStream#close()} is called.
+ */
+@SuppressWarnings("unused") // used via bytecode manipulation (SAGP)
+@Open
+public class SentryFileInputStream extends FileInputStream {
+
+  private final @Nullable ISpan currentSpan;
+  private final @Nullable File file;
+  private final @NotNull FileInputStream delegate;
+
+  private @NotNull SpanStatus spanStatus = SpanStatus.OK;
+  private long byteCount;
+
+  public SentryFileInputStream(@Nullable String name) throws FileNotFoundException {
+    this(init(name != null ? new File(name) : null, null));
+  }
+
+  public SentryFileInputStream(@Nullable File file) throws FileNotFoundException {
+    this(init(file, null));
+  }
+
+  public SentryFileInputStream(FileDescriptor fdObj) {
+    this(init(fdObj, null), fdObj);
+  }
+
+  private SentryFileInputStream(
+    @NotNull FileInputStreamInitData data,
+    @NotNull FileDescriptor fd
+  ) {
+    super(fd);
+    file = null;
+    currentSpan = data.span;
+    delegate = data.delegate;
+  }
+
+  private SentryFileInputStream(
+    @NotNull FileInputStreamInitData data
+  ) throws FileNotFoundException {
+    super(data.file);
+    currentSpan = data.span;
+    file = data.file;
+    delegate = data.delegate;
+  }
+
+  private static FileInputStreamInitData init(
+    @Nullable File file,
+    @Nullable FileInputStream delegate
+  ) throws FileNotFoundException {
+    final ISpan span = startSpan();
+    if (delegate == null) {
+      delegate = new FileInputStream(file);
+    }
+    return new FileInputStreamInitData(file, span, delegate);
+  }
+
+  private static FileInputStreamInitData init(
+    @NotNull FileDescriptor fd,
+    @Nullable FileInputStream delegate
+  ) {
+    final ISpan span = startSpan();
+    if (delegate == null) {
+      delegate = new FileInputStream(fd);
+    }
+    // TODO: it's only possible to get filename from FileDescriptor via reflection AND when it's
+    // TODO: running on Android, should we do that?
```

Comment:
```suggestion
    // TODO: it's only possible to get filename from FileDescriptor via reflection AND when it's
    // running on Android, should we do that?
```

Reply:
FileDescriptor is always used with FIOS right, and FIOS have file name, so we don't need to I guess

Reply:
hm, not sure I got you, FIOS have file name when it was instantiated with a filename or a File instance, but when the FileDescriptor ctor is used, then there's no access to the underlying File instance

Reply:
@romtsn tbh I don't recall using `FileDescriptor` so maybe what I've said indeed does not make sense, but looking into the Java docs and examples, eg https://www.javatpoint.com/java-filedescriptor-class
`FileDescriptor` is never used in the same way as standard File or FIOS, but always with it.

See `fd = fos.getFD(); `

in that case, we have access to file name because we'll be instrumenting the `fos` anyway

Reply:
Right, but the problem is that this can be returned from the Android SDK as a return type, and we do not instrument FIOS there. Anyway, I think it's fine for now, we will still instrument it, just w/o the filename/path

Reply:
then we are fine.